File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode();
        go(head, null, l1, l2, 0);
        return head;
    }

    public static void go(ListNode target, ListNode prev, ListNode l1, ListNode l2, int inMind) {
        if (l1 != null & l2 != null) {
            int sum = l1.val + l2.val + inMind;
            target.val = (sum > 9) ? sum % 10 : sum;
            target.next = (l1.next == null && l2.next == null) ? null : new ListNode();
            go(target.next, target, l1.next, l2.next, (sum > 9) ? 1 : 0);
        } else if (l1 != null) {
            int sum = l1.val + inMind;
            target.val = (sum > 9) ? sum % 10 : sum;
            target.next = (l1.next == null) ? null : new ListNode();
            go(target.next, target, l1.next, null, (sum > 9) ? 1 : 0);
        } else if (l2 != null) {
            int sum = l2.val + inMind;
            target.val = (sum > 9) ? sum % 10 : sum;
            target.next = (l2.next == null) ? null : new ListNode();
            go(target.next, target, null, l2.next, (sum > 9) ? 1 : 0);
        } else {
            if ((prev.val == 0 && inMind == 1) || inMind == 1) prev.next = new ListNode(1);
        }
    }

}

File: 0002-add-two-numbers/0002-add-two-numbers.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
    val res: ListNode = new ListNode()
    rec(l1, l2, 0, res, null)
    res
  }
    def rec(l1: ListNode, l2: ListNode, diff: Int, traversable: ListNode, prev: ListNode): Unit =
      (l1, l2) match {
        case (l1, l2) if l1 == null && l2 == null => if ((prev.x == 0 && diff == 1) || diff == 1) prev.next = new ListNode(1);
        case (l1, _) if l1 == null                =>
          val sum = l2.x + diff
          traversable.x = sum % 10
          traversable.next = if (l2.next == null) null else new ListNode()
          rec(l1, l2.next, (l2.x + diff) / 10, traversable.next, traversable)

        case (l1, l2) if l2 == null =>
          val sum = l1.x + diff
          traversable.x = sum  % 10
          traversable.next = if (l1.next == null) null else new ListNode()
          rec(l1.next, l2, sum / 10, traversable.next, traversable)

        case (l1, l2) =>
          val sum = l1.x + l2.x + diff
          traversable.x = sum % 10
          traversable.next = if (l1.next == null && l2.next == null) null else new ListNode()
          rec(l1.next, l2.next, sum / 10, traversable.next, traversable)
      }
}